File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution
{
public:
    int subarraySum(vector<int>& nums, int k)
    {
        map<int, int> mp;
        mp[0] = 1;                      // IMP { TC : [3, -3, 1, 1, 1] }
        int pfxSum = 0, answer = 0;

        for(int i = 0; i < nums.size(); i++)
        {
            pfxSum += nums[i];
            int remove = pfxSum - k;
            answer += mp[remove];
            mp[pfxSum]++;
        }

        return answer;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        res = 0
        curSum = 0
        pfxSums = {0 : 1}

        for n in nums:
            curSum += n
            diff = curSum - k
            res += pfxSums.get(diff, 0)
            pfxSums[curSum] = 1 + pfxSums.get(curSum, 0)

        return res